File: albedo-math/src/main/java/jp/albedo/jpl/state/impl/DirectStateSolver.java
package jp.albedo.jpl.state.impl;

import jp.albedo.common.RectangularCoordinates;
import jp.albedo.jpl.kernel.SpkKernelCollection;
import jp.albedo.jpl.state.StateSolver;
import jp.albedo.jpl.state.impl.chebyshev.PositionAndTrueVelocityCalculator;
import jp.albedo.jpl.state.impl.chebyshev.PositionAndVelocityCalculator;
import jp.albedo.jpl.state.impl.chebyshev.PositionAndVelocitySolvingCalculator;
import jp.albedo.utils.FunctionUtils;

import java.util.List;
import java.util.stream.Collectors;

public class DirectStateSolver implements StateSolver {

    final List<PositionAndVelocityCalculator> calculators;

    final boolean negate;

    public DirectStateSolver(List<SpkKernelCollection> spkKernelCollections, boolean negate) {
        this.calculators = spkKernelCollections.stream()
                .map(this::buildCalculator)
                .collect(Collectors.toList());
        this.negate = negate;
    }

    private PositionAndVelocityCalculator buildCalculator(SpkKernelCollection spkKernelCollection) {
        return spkKernelCollection.hasPositionAndVelocityData() ?
                new PositionAndTrueVelocityCalculator(spkKernelCollection.getPositionAndVelocityData()) :
                new PositionAndVelocitySolvingCalculator(spkKernelCollection.getPositionData());
    }

    public RectangularCoordinates positionFor(double ephemerisSeconds) {
        final RectangularCoordinates computed = calculators.stream()
                .map(FunctionUtils.wrap(calculator -> calculator.positionFor(ephemerisSeconds)))
                .reduce(RectangularCoordinates.ZERO, RectangularCoordinates::add);
        return negate ? computed.negate() : computed;
    }

    @Override
    public RectangularCoordinates velocityFor(double ephemerisSeconds) {
        final RectangularCoordinates computed = calculators.stream()
                .map(FunctionUtils.wrap(calculator -> calculator.velocityFor(ephemerisSeconds)))
                .reduce(RectangularCoordinates.ZERO, RectangularCoordinates::add);
        return negate ? computed.negate() : computed;
    }

}

File: api/math/jpl/state/DirectStateSolver.ts
import { RectangularCoordinates } from '../..';
import { SpkKernelCollection } from "../kernel";
import { StateSolver, PositionAndVelocityCalculator, PositionAndVelocitySolvingCalculator } from "./";

export class DirectStateSolver implements StateSolver {

    readonly calculators: PositionAndVelocityCalculator[];

    readonly negate: boolean;

    constructor(spkKernelCollections: SpkKernelCollection[], negate: boolean = false) {
        this.calculators = spkKernelCollections
            .map(this.buildCalculator);
        this.negate = negate;
    }

    buildCalculator(spkKernelCollection: SpkKernelCollection): PositionAndVelocityCalculator {
        return new PositionAndVelocitySolvingCalculator(spkKernelCollection.positionData);
        // return spkKernelCollection.hasPositionAndVelocityData() ?
        //         new PositionAndTrueVelocityCalculator(spkKernelCollection.getPositionAndVelocityData()) :
        //         new PositionAndVelocitySolvingCalculator(spkKernelCollection.getPositionData());
    }

    positionFor(ephemerisSeconds: number): RectangularCoordinates {
        const computed = this.calculators
            .map(calculator => calculator.positionFor(ephemerisSeconds))
            .reduce((a, b) => a.add(b), RectangularCoordinates.ZERO);
        return this.negate ? computed.negate() : computed;
    }

    velocityFor(ephemerisSeconds: number): RectangularCoordinates {
        const computed = this.calculators
            .map(calculator => calculator.velocityFor(ephemerisSeconds))
            .reduce((a, b) => a.add(b), RectangularCoordinates.ZERO);
        return this.negate ? computed.negate() : computed;
    }

}
